handlers/product-switch-api/src/frequencyChange.ts

```diff
@@ -0,0 +1,612 @@
+			error,
+		);
+		// Only return ValidationError messages to clients for security
+		if (error instanceof ValidationError) {
+			return {
+				reasons: [{ message: error.message }],
+			};
+		}
+		// Log unexpected errors but don't expose details to client
+		logger.log('Unexpected error type in frequency change processing', error);
+		return {
+			reasons: [
+				{
+					message: 'An unexpected error occurred while processing your request',
+				},
+			],
+		};
+	}
+}
+
+/**
+ * Frequency change handler
+ *
+ * @param stage Stage to execute in
+ * @param today Today's date
+ * @returns Http handler function for status code and body return
+ */
+export const frequencyChangeHandler =
+	(stage: Stage, today: dayjs.Dayjs) =>
+	async (
+		event: unknown,
+		parsed: {
+			path: { subscriptionNumber: string };
+			body: FrequencyChangeRequestBody;
+		},
+	): Promise<{ statusCode: number; body: string }> => {
+		logger.mutableAddContext(parsed.path.subscriptionNumber);
+		const todayDate = today.toDate();
+
+		const zuoraClient = await ZuoraClient.create(stage);
+		const subscription = await getSubscription(
+			zuoraClient,
+			parsed.path.subscriptionNumber,
+		);
+
+		// Extract identity ID from headers and validate ownership
+		const identityId = (
+			event as { headers?: Record<string, string | undefined> }
+		).headers?.['x-identity-id'];
+		const account = await getAccount(zuoraClient, subscription.accountNumber);
+
+		if (identityId && account.basicInfo.identityId !== identityId) {
+			logger.log(
+				`Subscription ${parsed.path.subscriptionNumber} does not belong to identity ID ${identityId}`,
+			);
+			return {
+				statusCode: 403,
+				body: JSON.stringify({
+					reasons: [
+						{
+							message: `Subscription ${parsed.path.subscriptionNumber} does not belong to the currently logged-in user`,
+						},
+					],
+				}),
+			};
+		}
+
+		const productCatalog = await getProductCatalogFromApi(stage);
+
+		// Use selectCandidateSubscriptionCharge to validate and find the eligible charge
+		let candidateCharge: { ratePlan: RatePlan; charge: RatePlanCharge };
+		try {
+			candidateCharge = selectCandidateSubscriptionCharge(
+				subscription,
+				todayDate,
+				account,
+			);
+		} catch (error) {
+			logger.log(
+				'Failed to select candidate charge for frequency change.',
+				error,
+			);
+			// Only return ValidationError messages to clients for security
+			if (error instanceof ValidationError) {
+				return {
+					statusCode: 400,
+					body: JSON.stringify({
+						reasons: [{ message: error.message }],
+					}),
+				};
+			}
+			// Log unexpected errors but don't expose details to client
+			logger.log('Unexpected error type in candidate charge selection', error);
+			return {
+				statusCode: 500,
+				body: JSON.stringify({
+					reasons: [
+						{
+							message:
+								'An unexpected error occurred while processing your request',
+						},
+					],
+				}),
+			};
```

Comment:
the router will do log-and-500 for free, to get the added context you can `throw new Error('Unexpected error type in candidate charge selection', {cause: e});`
```suggestion
			throw new Error('Unexpected error type in candidate charge selection', {cause: e});
```

Reply:
💪